agent/handlers/agentapi/taskprotection/v1/handlers/handlers.go

```diff
@@ -19,75 +19,187 @@ import (
+			Cluster:           aws.String(cluster),
+			ExpiresInMinutes:  taskProtection.GetExpiresInMinutes(),
+			ProtectionEnabled: aws.Bool(taskProtection.GetProtectionEnabled()),
+			Tasks:             aws.StringSlice([]string{task.Arn}),
+		})
+
+		logger.Debug("updateTaskProtection response:", logger.Fields{
+			loggerfield.TaskProtection: response.ProtectedTasks,
+			loggerfield.Error:          err,
+			loggerfield.Reason:         response.Failures,
+		})
+
+		if err != nil {
+			exceptionType, statusCode := getExceptionTypeAndStatusCode(err)
+			logger.Info(fmt.Sprintf("ECS throws an exception with type %s", exceptionType))
+			writeJSONResponse(w, statusCode, err.Error(), updateTaskProtectionRequestType)
+			return
+		}
+
+		// there are no exceptions but there are failures when setting protection in scheduler
+		if len(response.Failures) != 0 {
+			writeJSONResponse(w, http.StatusBadRequest, response.Failures[0], updateTaskProtectionRequestType)
+			return
+		}
+		if len(response.ProtectedTasks) != 1 {
+			msg := fmt.Sprintf("expect %v protectedTask in response, get %v", ExpectedProtectionResponseLength, len(response.ProtectedTasks))
+			logger.Error(msg)
+			writeJSONResponse(w, http.StatusInternalServerError, msg, updateTaskProtectionRequestType)
+
+			return
+		}
+		writeJSONResponse(w, http.StatusOK, response.ProtectedTasks[0], updateTaskProtectionRequestType)
+	}
+}
+
+// Helper function for retrieving credential from credentials manager and create ecs client
+func (factory TaskProtectionClientFactory) getECSClient(credentialsManager credentials.Manager, task *apitask.Task) (api.ECSTaskProtectionSDK, int, error) {
+	taskRoleCredential, ok := credentialsManager.GetTaskCredentials(task.GetCredentialsID())
+	if !ok {
+		return nil, http.StatusUnauthorized, errors.New("invalid Request: no task IAM role credentials available for task")
+	}
+	taskCredential := taskRoleCredential.GetIAMRoleCredentials()
+	cfg := aws.NewConfig().
+		WithCredentials(awscreds.NewStaticCredentials(taskCredential.AccessKeyID,
+			taskCredential.SecretAccessKey,
+			taskCredential.SessionToken)).
+		WithRegion(factory.Region).
+		WithHTTPClient(httpclient.New(ecsclient.RoundtripTimeout, factory.AcceptInsecureCert)).
+		WithEndpoint(factory.Endpoint)
+
+	ecsClient := ecs.New(session.Must(session.NewSession()), cfg)
+	return ecsClient, http.StatusOK, nil
+}
+
+// Helper function to parse exception type from error since it is unmarshalled in ecs client
+// see newClient method in agent/ecs_client/model/ecs/service.go
+func getExceptionTypeAndStatusCode(err error) (string, int) {
+	errorType := ""
+	// typical error format: InvalidParameterException: <Exception Message>
+	for i := range err.Error() {
+		if err.Error()[i] == ' ' {
+			errorType = err.Error()[0 : i-1] // trim colon
+			break
+		}
+	}
+
+	switch errorType {
+	case ecs.ErrCodeServerException:
+		return errorType, http.StatusInternalServerError
+	case ecs.ErrCodeAccessDeniedException,
```

Comment:
Regarding status code - I am using the status code specified by public ECS doc. There are no public ECS doc on TaskProtection APIs, will follow up adding a comment once it is ready.

https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_SubmitContainerStateChange.html#API_SubmitContainerStateChange_Errors